crates/executor/src/blockchain_tree/chain.rs

```diff
@@ -0,0 +1,319 @@
+            blocks: BTreeMap::from([(block.number, block.clone())]),
+        })
+    }
+
+    /// DONE
+    /// Create new chain that branches out from existing side chain.
+    pub fn new_chain_fork<SP: StateProvider, C: Consensus, EF: ExecutorFactory>(
+        &self,
+        block: SealedBlockWithSenders,
+        side_chain_block_hashes: BTreeMap<BlockNumber, BlockHash>,
+        canonical_block_hashes: &BTreeMap<BlockNumber, BlockHash>,
+        provider: &SP,
+        consensus: &C,
+        factory: &EF,
+    ) -> Result<Self, Error> {
+        let parent_nubmer = block.number - 1;
+        let parent = self
+            .blocks
+            .get(&parent_nubmer)
+            .ok_or(ExecError::BlockNumberNotFoundInChain { block_number: parent_nubmer })?;
+
+        // verify block against the parent
+        consensus.validate_header(&block, U256::ZERO)?;
+        consensus.pre_validate_header(&block, parent)?;
+        consensus.pre_validate_block(&block)?;
+
+        // revert changesets
+        let revert_from = self.changesets.len() - (self.tip().number - parent.number) as usize;
+        let mut substate = self.substate.clone();
+
+        // Revert changesets to get the state of the parent that we need to apply the change.
+        substate.revert(&self.changesets[revert_from..]);
+
+        let unseal = block.clone().into_components().0.unseal();
+
+        // execute block
+        let changeset = factory
+            .with_sp(SubStateWithProvider::new(
+                &substate,
+                provider,
+                &side_chain_block_hashes,
+                canonical_block_hashes,
+            ))
+            .execute_and_verify_receipt(&unseal, U256::MAX, None)?;
+
+        let chain = Self {
+            substate,
+            changesets: vec![changeset],
+            blocks: BTreeMap::from([(block.number, block)]),
+        };
+
+        // if all is okay, return new chain back. Present chain is not modified.
+        Ok(chain)
+    }
+
+    /// Append block to this chain
+    pub fn append_block<SP: StateProvider, C: Consensus, EF: ExecutorFactory>(
+        &mut self,
+        block: SealedBlockWithSenders,
+        side_chain_block_hashes: BTreeMap<BlockNumber, BlockHash>,
+        canonical_block_hashes: &BTreeMap<BlockNumber, BlockHash>,
+        provider: &SP,
+        consensus: &C,
+        factory: &EF,
+    ) -> Result<(), Error> {
+        let (_, parent) = self.blocks.last_key_value().expect("Chain has at least one block");
+
+        consensus.validate_header(&block, U256::ZERO)?;
+        consensus.pre_validate_header(&block, parent)?;
+        consensus.pre_validate_block(&block)?;
+
+        let unseal = block.clone().into_components().0.unseal();
+        let changeset = factory
+            .with_sp(SubStateWithProvider::new(
+                &self.substate,
+                provider,
+                &side_chain_block_hashes,
+                canonical_block_hashes,
+            ))
```

Comment:
All of this is duplicated 3 times, please consider consolidating into an internal function so its easier to change

Reply:
They are slightly different, but wrapping them in function would indeed clean up some things